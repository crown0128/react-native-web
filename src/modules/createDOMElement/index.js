import '../injectResponderEventPlugin';

import getAccessibilityRole from '../getAccessibilityRole';
import normalizeNativeEvent from '../normalizeNativeEvent';
import React from 'react';
import StyleRegistry from '../../apis/StyleSheet/registry';

const emptyObject = {};

const roleComponents = {
  article: 'article',
  banner: 'header',
  button: 'button',
  complementary: 'aside',
  contentinfo: 'footer',
  form: 'form',
  heading: 'h1',
  link: 'a',
  list: 'ul',
  listitem: 'li',
  main: 'main',
  navigation: 'nav',
  region: 'section'
};

const eventHandlerNames = {
  onClick: true,
  onClickCapture: true,
  onMoveShouldSetResponder: true,
  onMoveShouldSetResponderCapture: true,
  onResponderGrant: true,
  onResponderMove: true,
  onResponderReject: true,
  onResponderRelease: true,
  onResponderTerminate: true,
  onResponderTerminationRequest: true,
  onStartShouldSetResponder: true,
  onStartShouldSetResponderCapture: true,
  onTouchCancel: true,
  onTouchCancelCapture: true,
  onTouchEnd: true,
  onTouchEndCapture: true,
  onTouchMove: true,
  onTouchMoveCapture: true,
  onTouchStart: true,
  onTouchStartCapture: true
};

const wrapEventHandler = handler => e => {
  e.nativeEvent = normalizeNativeEvent(e.nativeEvent);
  return handler(e);
};

const createDOMElement = (component, rnProps) => {
  const {
    accessibilityLabel,
    accessibilityLiveRegion,
    accessible = true,
    style: rnStyle,
    testID,
    type,
    /* eslint-disable */
    accessibilityComponentType,
    accessibilityRole,
    accessibilityTraits,
    /* eslint-enable */
    ...domProps
  } = rnProps || emptyObject;

  // use equivalent platform elements where possible
  const role = getAccessibilityRole(rnProps || emptyObject);
  const accessibilityComponent = role && roleComponents[role];
  const Component = accessibilityComponent || component;

  // convert React Native styles to DOM styles
  const { className, style } = StyleRegistry.resolve(rnStyle) || emptyObject;

  // normalize DOM events to match React Native events
  // TODO: move this out of the render path
  for (const prop in domProps) {
    if (Object.prototype.hasOwnProperty.call(domProps, prop)) {
      const isEventHandler = typeof prop === 'function' && eventHandlerNames[prop];
      if (isEventHandler) {
        domProps[prop] = wrapEventHandler(prop);
      }
    }
  }

  if (!accessible) {
    domProps['aria-hidden'] = true;
  }
  if (accessibilityLabel) {
    domProps['aria-label'] = accessibilityLabel;
  }
  if (accessibilityLiveRegion) {
    domProps['aria-live'] = accessibilityLiveRegion;
  }
  if (className && className !== '') {
    domProps.className = domProps.className ? `${domProps.className} ${className}` : className;
  }
  if (role) {
    domProps.role = role;
    if (role === 'button') {
      domProps.type = 'button';
    } else if (role === 'link' && domProps.target === '_blank') {
      domProps.rel = `${domProps.rel || ''} noopener noreferrer`;
    }
  }
  if (style) {
    domProps.style = style;
  }
  if (testID) {
    domProps['data-testid'] = testID;
  }
  if (type) {
    domProps.type = type;
  }

  return <Component {...domProps} />;
};

module.exports = createDOMElement;
