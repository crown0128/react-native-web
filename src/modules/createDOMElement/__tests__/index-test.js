/* eslint-env mocha */

import assert from 'assert'
import createDOMElement from '..'
import { shallow } from 'enzyme'

suite('modules/createDOMElement', () => {
  test('renders correct DOM element', () => {
    let element = shallow(createDOMElement('span'))
    assert.equal(element.is('span'), true)
    element = shallow(createDOMElement('main'))
    assert.equal(element.is('main'), true)
  })

  test('prop "accessibilityLabel"', () => {
    const accessibilityLabel = 'accessibilityLabel'
    const element = shallow(createDOMElement('span', { accessibilityLabel }))
    assert.equal(element.prop('aria-label'), accessibilityLabel)
  })

  test('prop "accessibilityLiveRegion"', () => {
    const accessibilityLiveRegion = 'polite'
    const element = shallow(createDOMElement('span', { accessibilityLiveRegion }))
    assert.equal(element.prop('aria-live'), accessibilityLiveRegion)
  })

  test('prop "accessibilityRole"', () => {
    const accessibilityRole = 'banner'
    let element = shallow(createDOMElement('span', { accessibilityRole }))
    assert.equal(element.prop('role'), accessibilityRole)
    assert.equal(element.is('header'), true)

    const button = 'button'
    element = shallow(createDOMElement('span', { accessibilityRole: 'button' }))
    assert.equal(element.prop('type'), button)
    assert.equal(element.is('button'), true)
  })

  test('prop "accessible"', () => {
    // accessible (implicit)
    let element = shallow(createDOMElement('span', {}))
    assert.equal(element.prop('aria-hidden'), null)
    // accessible (explicit)
    element = shallow(createDOMElement('span', { accessible: true }))
    assert.equal(element.prop('aria-hidden'), null)
    // not accessible
    element = shallow(createDOMElement('span', { accessible: false }))
    assert.equal(element.prop('aria-hidden'), true)
  })

  test('prop "testID"', () => {
    // no testID
    let element = shallow(createDOMElement('span', {}))
    assert.equal(element.prop('data-testid'), null)
    // with testID
    const testID = 'Example.testID'
    element = shallow(createDOMElement('span', { testID }))
    assert.equal(element.prop('data-testid'), testID)
  })
})
