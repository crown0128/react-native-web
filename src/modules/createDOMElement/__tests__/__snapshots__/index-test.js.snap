exports[`modules/createDOMElement prop "accessibilityLabel" 1`] = `
<span
  aria-label="accessibilityLabel" />
`;

exports[`modules/createDOMElement prop "accessibilityLiveRegion" 1`] = `
<span
  aria-live="polite" />
`;

exports[`modules/createDOMElement prop "accessibilityRole" button 1`] = `
<button
  role="button"
  type="button" />
`;

exports[`modules/createDOMElement prop "accessibilityRole" compatibility with accessibilityComponentType 1`] = `
<button
  role="button"
  type="button" />
`;

exports[`modules/createDOMElement prop "accessibilityRole" compatibility with accessibilityComponentType 2`] = `
<a
  role="link" />
`;

exports[`modules/createDOMElement prop "accessibilityRole" compatibility with accessibilityTraits 1`] = `
<button
  role="button"
  type="button" />
`;

exports[`modules/createDOMElement prop "accessibilityRole" compatibility with accessibilityTraits 2`] = `
<a
  role="link" />
`;

exports[`modules/createDOMElement prop "accessibilityRole" link and target="_blank" 1`] = `
<a
  rel=" noopener noreferrer"
  role="link"
  target="_blank" />
`;

exports[`modules/createDOMElement prop "accessibilityRole" roles 1`] = `
<header
  role="banner" />
`;

exports[`modules/createDOMElement prop "accessible" 1`] = `<span />`;

exports[`modules/createDOMElement prop "accessible" 2`] = `<span />`;

exports[`modules/createDOMElement prop "accessible" 3`] = `
<span
  aria-hidden={true} />
`;

exports[`modules/createDOMElement prop "testID" 1`] = `
<span
  data-testid="Example.testID" />
`;

exports[`modules/createDOMElement renders correct DOM element 1`] = `<span />`;

exports[`modules/createDOMElement renders correct DOM element 2`] = `<main />`;
